File: scene/img/process-assets.js
import sharp from 'sharp';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const inputDir = __dirname;
const outputDir = __dirname;

// 配置
const config = {
  headshot: {
    sizes: [64, 128, 256],
    suffixes: ['@1x', '@2x', '@4x'],
    quality: 85,
    format: 'png'
  }
};

/**
 * 创建圆形 SVG 遮罩
 */
function createCircleMask(size) {
  return `
    <svg width="${size}" height="${size}" xmlns="http://www.w3.org/2000/svg">
      <defs>
        <mask id="circle">
          <rect width="${size}" height="${size}" fill="white"/>
          <circle cx="${size/2}" cy="${size/2}" r="${size/2 - 1}" fill="black"/>
        </mask>
      </defs>
      <circle cx="${size/2}" cy="${size/2}" r="${size/2 - 1}" fill="white" mask="url(#circle)"/>
    </svg>
  `;
}

/**
 * 生成游戏可用的头像（多分辨率）
 */
async function generateHeadshotVariants(inputPath) {
  const baseName = path.basename(inputPath, path.extname(inputPath));
  const sizes = config.headshot.sizes;
  const suffixes = config.headshot.suffixes;

  try {
    console.log(`\n🎨 处理头像: ${baseName}`);
    console.log(`${'─'.repeat(50)}`);

    // 首先获取图像元数据并做初步处理
    const metadata = await sharp(inputPath).metadata();
    const minDim = Math.min(metadata.width || 0, metadata.height || 0);

    if (minDim === 0) {
      console.error(`❌ 无法读取图像尺寸`);
      return false;
    }

    // 裁剪为正方形（优先裁剪中心，保留脸部）
    const crop = Math.floor((Math.max(metadata.width || 0, metadata.height || 0) - minDim) / 2);
    const squareImage = await sharp(inputPath)
      .extract({ 
        left: crop, 
        top: crop, 
        width: minDim, 
        height: minDim 
      })
      .toBuffer();

    let processed = 0;

    for (let i = 0; i < sizes.length; i++) {
      const size = sizes[i];
      const suffix = suffixes[i];
      const outputPath = path.join(outputDir, `${baseName}-head${suffix}.png`);

      try {
        // 调整大小并应用圆形遮罩
        await sharp(squareImage)
          .resize(size, size, { fit: 'cover', position: 'center' })
          .composite([
            {
              input: Buffer.from(createCircleMask(size)),
              blend: 'dest-in'
            }
          ])
          .png({ quality: config.headshot.quality })
          .toFile(outputPath);

        console.log(`   ✓ ${baseName}-head${suffix}.png (${size}×${size}px)`);
        processed++;
      } catch (error) {
        console.error(`   ✗ 生成 ${suffix} 版本失败: ${error.message}`);
      }
    }

    if (processed > 0) {
      console.log(`\n✅ 成功生成 ${processed} 个头像版本`);
      return true;
    }
    return false;

  } catch (error) {
    console.error(`❌ 工 处理失败: ${error.message}`);
    return false;
  }
}

/**
 * 列出输入目录中的所有图像
 */
function findImages(dir) {
  const supportedFormats = ['.jpg', '.jpeg', '.png', '.webp', '.bmp'];

  try {
    const files = fs.readdirSync(dir);
    return files.filter(file => {
      const ext = path.extname(file).toLowerCase();
      const name = path.basename(file);
      return (
        supportedFormats.includes(ext) &&
        !name.startsWith('_') &&
        !name.startsWith('.') &&
        !name.includes('-head@') &&
        !name.includes('-character')
      );
    });
  } catch (error) {
    return [];
  }
}

/**
 * 主处理函数
 */
async function main() {
  const args = process.argv.slice(2);
  const typeArg = args.find(a => a.startsWith('--type'))?.split('=')[1] || 'headshot';

  console.log(`\n${'═'.repeat(50)}`);
  console.log(`🎮 医院游戏素材处理工程`);
  console.log(`═'.repeat(50)}\n`);

  const images = findImages(inputDir);

  if (images.length === 0) {
    console.log(`⚠️  未找到图像文件\n`);
    console.log(`📋 步骤：`);
    console.log(`   1. 将你的大头照放在 img/ 目录中`);
    console.log(`   2. 支持格式: JPG, PNG, WebP, BMP`);
    console.log(`   3. 运行: npm run process\n`);
    console.log(`📝 示例：`);
    console.log(`   · 将 photo.jpg 放在 img/ 目录`);
    console.log(`   · 运行 npm run process`);
    console.log(`   · 得到 photo-head@1x.png 等文件\n`);
    console.log(`${'─'.repeat(50)}\n`);
    return;
  }

  console.log(`发现 ${images.length} 个图像文件:`);
  images.forEach((img, i) => console.log(`   ${i + 1}. ${img}`));
  console.log();

  let successCount = 0;
  let errorCount = 0;

  for (const image of images) {
    const inputPath = path.join(inputDir, image);

    if (typeArg === 'headshot' || typeArg === 'all') {
      const success = await generateHeadshotVariants(inputPath);
      if (success) successCount++;
      else errorCount++;
    }
  }

  console.log(`\n${'═'.repeat(50)}`);
  console.log(`📊 处理完成`);
  console.log(`${'─'.repeat(50)}`);
  console.log(`✅ 成功: ${successCount} 个`);
  console.log(`❌ 失败: ${errorCount} 个`);
  console.log(`📍 输出位置: ${outputDir}`);
  console.log(`\n💡 后续步骤:`);
  console.log(`   1. 检查生成的 -head@2x.png 文件`);
  console.log(`   2. 仔细阅读 img/README.md`);
  console.log(`   3. 按说明集成到游戏代码\n`);
  console.log(`${'═'.repeat(50)}\n`);
}

main().catch(err => {
  console.error(`\n❌ 致命错误: ${err.message}\n`);
  process.exit(1);
});

File: co_op_frontend/img/process-assets.py
"""
医院游戏素材处理工程 - Python 版本
将大头照处理成游戏可用的圆形头像（多分辨率）
"""

import os
import sys
from pathlib import Path
from PIL import Image, ImageDraw
import argparse

# 配置
SCRIPT_DIR = Path(__file__).parent
SUPPORTED_FORMATS = {'.jpg', '.jpeg', '.png', '.webp', '.bmp', '.tiff', '.gif'}

HEADSHOT_CONFIG = {
    'sizes': [64, 128, 256],
    'suffixes': ['@1x', '@2x', '@4x'],
    'quality': 85
}


def create_circle_mask(size):
    """创建圆形遮罩"""
    mask = Image.new('L', (size, size), 0)
    draw = ImageDraw.Draw(mask)
    draw.ellipse([0, 0, size - 1, size - 1], fill=255)
    return mask


def process_headshot(input_path):
    """
    处理大头照 -> 圆形头像（多分辨率）
    
    Args:
        input_path: 输入图像路径
        
    Returns:
        bool: 处理是否成功
    """
    base_name = Path(input_path).stem
    
    print(f"\n🎨 处理头像: {base_name}")
    print("─" * 50)
    
    try:
        # 打开原始图像
        img = Image.open(input_path).convert('RGBA')
        width, height = img.size
        
        # 裁剪为正方形（保留中心，优先保留脸部）
        min_dim = min(width, height)
        left = (width - min_dim) // 2
        top = (height - min_dim) // 2
        right = left + min_dim
        bottom = top + min_dim
        
        square_img = img.crop((left, top, right, bottom))
        
        processed_count = 0
        
        for i, size in enumerate(HEADSHOT_CONFIG['sizes']):
            suffix = HEADSHOT_CONFIG['suffixes'][i]
            output_name = f"{base_name}-head{suffix}.png"
            output_path = SCRIPT_DIR / output_name
            
            try:
                # 调整大小
                resized = square_img.resize((size, size), Image.Resampling.LANCZOS)
                
                # 创建空白背景（透明）
                result = Image.new('RGBA', (size, size), (0, 0, 0, 0))
                
                # 应用圆形遮罩
                mask = create_circle_mask(size)
                result.paste(resized, (0, 0), mask)
                
                # 保存
                result.save(
                    output_path,
                    'PNG',
                    optimize=False,
                    quality=HEADSHOT_CONFIG['quality']
                )
                
                print(f"   ✓ {output_name} ({size}×{size}px)")
                processed_count += 1
                
            except Exception as e:
                print(f"   ✗ 生成 {suffix} 版本失败: {e}")
        
        if processed_count > 0:
            print(f"\n✅ 成功生成 {processed_count} 个头像版本")
            return True
        
        return False
        
    except Exception as e:
        print(f"❌ 处理失败: {e}")
        return False


def find_images(directory):
    """列出目录中的所有未处理的图像文件"""
    images = []
    
    for file in Path(directory).iterdir():
        if not file.is_file():
            continue
            
        ext = file.suffix.lower()
        name = file.name
        
        # 跳过已处理的文件、隐藏文件和系统文件
        if (ext in SUPPORTED_FORMATS and 
            not name.startswith('_') and 
            not name.startswith('.') and
            '-head@' not in name and
            '-character' not in name):
            images.append(file.name)
    
    return sorted(images)


def main():
    parser = argparse.ArgumentParser(
        description='医院游戏素材处理工程',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
示例:
  python process-assets.py              # 处理所有图像为头像
  python process-assets.py --type all   # 全量处理
        """
    )
    parser.add_argument(
        '--type',
        default='headshot',
        choices=['headshot', 'all'],
        help='处理类型（默认: headshot）'
    )
    
    args = parser.parse_args()
    
    print("\n" + "═" * 50)
    print("🎮 医院游戏素材处理工程 (Python 版)")
    print("═" * 50 + "\n")
    
    images = find_images(SCRIPT_DIR)
    
    if not images:
        print("⚠️  未找到图像文件\n")
        print("📋 步骤：")
        print("   1. 将你的大头照放在 img/ 目录中")
        print(f"   2. 支持格式: {', '.join(SUPPORTED_FORMATS)}")
        print("   3. 运行: python process-assets.py\n")
        print("📝 示例：")
        print("   · 将 photo.jpg 放在 img/ 目录")
        print("   · 运行 python process-assets.py")
        print("   · 得到 photo-head@1x.png 等文件\n")
        print("─" * 50 + "\n")
        return
    
    print(f"发现 {len(images)} 个图像文件:")
    for i, img in enumerate(images, 1):
        print(f"   {i}. {img}")
    print()
    
    success_count = 0
    error_count = 0
    
    for image in images:
        input_path = SCRIPT_DIR / image
        
        if args.type in ['headshot', 'all']:
            if process_headshot(input_path):
                success_count += 1
            else:
                error_count += 1
    
    print("\n" + "═" * 50)
    print("📊 处理完成")
    print("─" * 50)
    print(f"✅ 成功: {success_count} 个")
    if error_count > 0:
        print(f"❌ 失败: {error_count} 个")
    print(f"📍 输出位置: {SCRIPT_DIR}")
    print("\n💡 后续步骤:")
    print("   1. 检查生成的 -head@2x.png 文件")
    print("   2. 仔细阅读 img/README.md")
    print("   3. 按说明集成到游戏代码")
    print("\n" + "═" * 50 + "\n")


if __name__ == '__main__':
    main()
